In [1]:
##Import relevant libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
import seaborn as sns

In [2]:
raw_data = pd.read_csv("C:/Users/Administrator/Downloads\Heart.csv")

In [3]:
raw_data

Unnamed: 0  Age  Sex     ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  \
0             1   63    1       typical     145   233    1        2    150   
1             2   67    1  asymptomatic     160   286    0        2    108   
2             3   67    1  asymptomatic     120   229    0        2    129   
3             4   37    1    nonanginal     130   250    0        0    187   
4             5   41    0    nontypical     130   204    0        2    172   
..          ...  ...  ...           ...     ...   ...  ...      ...    ...   
298         299   45    1       typical     110   264    0        0    132   
299         300   68    1  asymptomatic     144   193    1        0    141   
300         301   57    1  asymptomatic     130   131    0        0    115   
301         302   57    0    nontypical     130   236    0        2    174   
302         303   38    1    nonanginal     138   175    0        0    173   

     ExAng  Oldpeak  Slope   Ca        Thal  AHD  
0        0      2.3      3  0.0       fixed   No  
1        1      1.5      2  3.0      normal  Yes  
2        1      2.6      2  2.0  reversable  Yes  
3        0      3.5      3  0.0      normal   No  
4        0      1.4      1  0.0      normal   No  
..     ...      ...    ...  ...         ...  ...  
298      0      1.2      2  0.0  reversable  Yes  
299      0      3.4      2  2.0  reversable  Yes  
300      1      1.2      2  1.0  reversable  Yes  
301      0      0.0      2  1.0      normal  Yes  
302      0      0.0      1  NaN      normal   No  

[303 rows x 15 columns]

In [4]:
raw_data.describe()

Unnamed: 0         Age         Sex      RestBP        Chol         Fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean   152.000000   54.438944    0.679868  131.689769  246.693069    0.148515   
std     87.612784    9.038662    0.467299   17.599748   51.776918    0.356198   
min      1.000000   29.000000    0.000000   94.000000  126.000000    0.000000   
25%     76.500000   48.000000    0.000000  120.000000  211.000000    0.000000   
50%    152.000000   56.000000    1.000000  130.000000  241.000000    0.000000   
75%    227.500000   61.000000    1.000000  140.000000  275.000000    0.000000   
max    303.000000   77.000000    1.000000  200.000000  564.000000    1.000000   

          RestECG       MaxHR       ExAng     Oldpeak       Slope          Ca  
count  303.000000  303.000000  303.000000  303.000000  303.000000  299.000000  
mean     0.990099  149.607261    0.326733    1.039604    1.600660    0.672241  
std      0.994971   22.875003    0.469794    1.161075    0.616226    0.937438  
min      0.000000   71.000000    0.000000    0.000000    1.000000    0.000000  
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000  
50%      1.000000  153.000000    0.000000    0.800000    2.000000    0.000000  
75%      2.000000  166.000000    1.000000    1.600000    2.000000    1.000000  
max      2.000000  202.000000    1.000000    6.200000    3.000000    3.000000

In [5]:
raw_data = raw_data.drop(columns = 'Unnamed: 0')
raw_data

Age  Sex     ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  \
0     63    1       typical     145   233    1        2    150      0   
1     67    1  asymptomatic     160   286    0        2    108      1   
2     67    1  asymptomatic     120   229    0        2    129      1   
3     37    1    nonanginal     130   250    0        0    187      0   
4     41    0    nontypical     130   204    0        2    172      0   
..   ...  ...           ...     ...   ...  ...      ...    ...    ...   
298   45    1       typical     110   264    0        0    132      0   
299   68    1  asymptomatic     144   193    1        0    141      0   
300   57    1  asymptomatic     130   131    0        0    115      1   
301   57    0    nontypical     130   236    0        2    174      0   
302   38    1    nonanginal     138   175    0        0    173      0   

     Oldpeak  Slope   Ca        Thal  AHD  
0        2.3      3  0.0       fixed   No  
1        1.5      2  3.0      normal  Yes  
2        2.6      2  2.0  reversable  Yes  
3        3.5      3  0.0      normal   No  
4        1.4      1  0.0      normal   No  
..       ...    ...  ...         ...  ...  
298      1.2      2  0.0  reversable  Yes  
299      3.4      2  2.0  reversable  Yes  
300      1.2      2  1.0  reversable  Yes  
301      0.0      2  1.0      normal  Yes  
302      0.0      1  NaN      normal   No  

[303 rows x 14 columns]

In [7]:
## One-hot encode categorical columns and ensure 0/1 output
raw_data = pd.get_dummies(raw_data, columns=['ChestPain', 'Thal'], drop_first=True, dtype=int)

##Encode target variable (AHD) as 0 and 1
raw_data['AHD'] = raw_data['AHD'].map({'Yes': 1, 'No': 0})


In [8]:
raw_data

Age  Sex  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  Oldpeak  Slope   Ca  \
0     63    1     145   233    1        2    150      0      2.3      3  0.0   
1     67    1     160   286    0        2    108      1      1.5      2  3.0   
2     67    1     120   229    0        2    129      1      2.6      2  2.0   
3     37    1     130   250    0        0    187      0      3.5      3  0.0   
4     41    0     130   204    0        2    172      0      1.4      1  0.0   
..   ...  ...     ...   ...  ...      ...    ...    ...      ...    ...  ...   
298   45    1     110   264    0        0    132      0      1.2      2  0.0   
299   68    1     144   193    1        0    141      0      3.4      2  2.0   
300   57    1     130   131    0        0    115      1      1.2      2  1.0   
301   57    0     130   236    0        2    174      0      0.0      2  1.0   
302   38    1     138   175    0        0    173      0      0.0      1  NaN   

     AHD  ChestPain_nonanginal  ChestPain_nontypical  ChestPain_typical  \
0      0                     0                     0                  1   
1      1                     0                     0                  0   
2      1                     0                     0                  0   
3      0                     1                     0                  0   
4      0                     0                     1                  0   
..   ...                   ...                   ...                ...   
298    1                     0                     0                  1   
299    1                     0                     0                  0   
300    1                     0                     0                  0   
301    1                     0                     1                  0   
302    0                     1                     0                  0   

     Thal_normal  Thal_reversable  
0              0                0  
1              1                0  
2              0                1  
3              1                0  
4              1                0  
..           ...              ...  
298            0                1  
299            0                1  
300            0                1  
301            1                0  
302            1                0  

[303 rows x 17 columns]

In [9]:
raw_data.isnull().sum()

Age                     0
Sex                     0
RestBP                  0
Chol                    0
Fbs                     0
RestECG                 0
MaxHR                   0
ExAng                   0
Oldpeak                 0
Slope                   0
Ca                      4
AHD                     0
ChestPain_nonanginal    0
ChestPain_nontypical    0
ChestPain_typical       0
Thal_normal             0
Thal_reversable         0
dtype: int64

In [10]:
data = raw_data.dropna()
data

Age  Sex  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  Oldpeak  Slope   Ca  \
0     63    1     145   233    1        2    150      0      2.3      3  0.0   
1     67    1     160   286    0        2    108      1      1.5      2  3.0   
2     67    1     120   229    0        2    129      1      2.6      2  2.0   
3     37    1     130   250    0        0    187      0      3.5      3  0.0   
4     41    0     130   204    0        2    172      0      1.4      1  0.0   
..   ...  ...     ...   ...  ...      ...    ...    ...      ...    ...  ...   
297   57    0     140   241    0        0    123      1      0.2      2  0.0   
298   45    1     110   264    0        0    132      0      1.2      2  0.0   
299   68    1     144   193    1        0    141      0      3.4      2  2.0   
300   57    1     130   131    0        0    115      1      1.2      2  1.0   
301   57    0     130   236    0        2    174      0      0.0      2  1.0   

     AHD  ChestPain_nonanginal  ChestPain_nontypical  ChestPain_typical  \
0      0                     0                     0                  1   
1      1                     0                     0                  0   
2      1                     0                     0                  0   
3      0                     1                     0                  0   
4      0                     0                     1                  0   
..   ...                   ...                   ...                ...   
297    1                     0                     0                  0   
298    1                     0                     0                  1   
299    1                     0                     0                  0   
300    1                     0                     0                  0   
301    1                     0                     1                  0   

     Thal_normal  Thal_reversable  
0              0                0  
1              1                0  
2              0                1  
3              1                0  
4              1                0  
..           ...              ...  
297            0                1  
298            0                1  
299            0                1  
300            0                1  
301            1                0  

[299 rows x 17 columns]

In [11]:
x = data.drop(columns = 'AHD')
x

Age  Sex  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  Oldpeak  Slope   Ca  \
0     63    1     145   233    1        2    150      0      2.3      3  0.0   
1     67    1     160   286    0        2    108      1      1.5      2  3.0   
2     67    1     120   229    0        2    129      1      2.6      2  2.0   
3     37    1     130   250    0        0    187      0      3.5      3  0.0   
4     41    0     130   204    0        2    172      0      1.4      1  0.0   
..   ...  ...     ...   ...  ...      ...    ...    ...      ...    ...  ...   
297   57    0     140   241    0        0    123      1      0.2      2  0.0   
298   45    1     110   264    0        0    132      0      1.2      2  0.0   
299   68    1     144   193    1        0    141      0      3.4      2  2.0   
300   57    1     130   131    0        0    115      1      1.2      2  1.0   
301   57    0     130   236    0        2    174      0      0.0      2  1.0   

     ChestPain_nonanginal  ChestPain_nontypical  ChestPain_typical  \
0                       0                     0                  1   
1                       0                     0                  0   
2                       0                     0                  0   
3                       1                     0                  0   
4                       0                     1                  0   
..                    ...                   ...                ...   
297                     0                     0                  0   
298                     0                     0                  1   
299                     0                     0                  0   
300                     0                     0                  0   
301                     0                     1                  0   

     Thal_normal  Thal_reversable  
0              0                0  
1              1                0  
2              0                1  
3              1                0  
4              1                0  
..           ...              ...  
297            0                1  
298            0                1  
299            0                1  
300            0                1  
301            1                0  

[299 rows x 16 columns]

In [12]:
y = data['AHD']

In [14]:
from sklearn.model_selection import train_test_split

In [15]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 365)

In [16]:
x_train

Age  Sex  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  Oldpeak  Slope   Ca  \
246   58    1     100   234    0        0    156      0      0.1      1  1.0   
111   56    1     125   249    1        2    144      1      1.2      2  1.0   
250   57    1     110   201    0        0    126      1      1.5      2  0.0   
0     63    1     145   233    1        2    150      0      2.3      3  0.0   
225   34    0     118   210    0        0    192      0      0.7      1  0.0   
..   ...  ...     ...   ...  ...      ...    ...    ...      ...    ...  ...   
228   54    1     110   206    0        2    108      1      0.0      2  1.0   
263   44    1     120   226    0        0    169      0      0.0      1  0.0   
292   44    1     120   169    0        0    144      1      2.8      3  0.0   
181   56    0     134   409    0        2    150      1      1.9      2  2.0   
82    39    1     140   321    0        2    182      0      0.0      1  0.0   

     ChestPain_nonanginal  ChestPain_nontypical  ChestPain_typical  \
246                     0                     0                  0   
111                     0                     0                  0   
250                     0                     0                  0   
0                       0                     0                  1   
225                     0                     1                  0   
..                    ...                   ...                ...   
228                     0                     0                  0   
263                     1                     0                  0   
292                     0                     0                  0   
181                     0                     0                  0   
82                      1                     0                  0   

     Thal_normal  Thal_reversable  
246            0                1  
111            1                0  
250            0                0  
0              0                0  
225            1                0  
..           ...              ...  
228            1                0  
263            1                0  
292            0                0  
181            0                1  
82             1                0  

[209 rows x 16 columns]

In [17]:
x_test

Age  Sex  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  Oldpeak  Slope   Ca  \
85    44    1     140   235    0        2    180      0      0.0      1  0.0   
234   54    0     160   201    0        0    163      0      0.0      1  1.0   
20    64    1     110   211    0        2    144      1      1.8      2  0.0   
123   55    1     140   217    0        0    111      1      5.6      3  0.0   
295   41    1     120   157    0        0    182      0      0.0      1  0.0   
..   ...  ...     ...   ...  ...      ...    ...    ...      ...    ...  ...   
59    51    1     125   213    0        2    125      1      1.4      1  1.0   
136   70    1     145   174    0        0    125      1      2.6      3  0.0   
130   54    1     120   258    0        2    147      0      0.4      2  0.0   
96    59    1     110   239    0        2    142      1      1.2      2  1.0   
68    59    1     170   326    0        2    140      1      3.4      3  0.0   

     ChestPain_nonanginal  ChestPain_nontypical  ChestPain_typical  \
85                      1                     0                  0   
234                     1                     0                  0   
20                      0                     0                  1   
123                     0                     0                  0   
295                     0                     1                  0   
..                    ...                   ...                ...   
59                      0                     0                  1   
136                     0                     0                  0   
130                     1                     0                  0   
96                      0                     0                  0   
68                      0                     0                  0   

     Thal_normal  Thal_reversable  
85             1                0  
234            1                0  
20             1                0  
123            0                1  
295            1                0  
..           ...              ...  
59             1                0  
136            0                1  
130            0                1  
96             0                1  
68             0                1  

[90 rows x 16 columns]

In [22]:
##Standardize the data

from sklearn.preprocessing import StandardScaler

In [23]:
scaler = StandardScaler()

In [24]:
x_train_scaled = scaler.fit_transform(x_train)

In [25]:
x_test_scaled = scaler.transform(x_test)

In [26]:
x_train_scaled

array([[ 0.38246661,  0.72494651, -1.75695006, ..., -0.25753938,
        -1.06426174,  1.23210076],
       [ 0.16225856,  0.72494651, -0.36392061, ..., -0.25753938,
         0.93961848, -0.81162193],
       [ 0.27236258,  0.72494651, -1.19973828, ..., -0.25753938,
        -1.06426174, -0.81162193],
       ...,
       [-1.15898971,  0.72494651, -0.6425265 , ..., -0.25753938,
        -1.06426174, -0.81162193],
       [ 0.16225856, -1.37941211,  0.13756999, ..., -0.25753938,
        -1.06426174,  1.23210076],
       [-1.70950983,  0.72494651,  0.47189706, ..., -0.25753938,
         0.93961848, -0.81162193]])

In [27]:
x_test_scaled

array([[-1.15898971,  0.72494651,  0.47189706, ..., -0.25753938,
         0.93961848, -0.81162193],
       [-0.05794949, -1.37941211,  1.58632062, ..., -0.25753938,
         0.93961848, -0.81162193],
       [ 1.04309074,  0.72494651, -1.19973828, ...,  3.88290137,
         0.93961848, -0.81162193],
       ...,
       [-0.05794949,  0.72494651, -0.6425265 , ..., -0.25753938,
        -1.06426174,  1.23210076],
       [ 0.49257063,  0.72494651, -1.19973828, ..., -0.25753938,
        -1.06426174,  1.23210076],
       [ 0.49257063,  0.72494651,  2.1435324 , ..., -0.25753938,
        -1.06426174,  1.23210076]])

In [28]:
from sklearn.linear_model import LogisticRegression

In [29]:
log_reg = LogisticRegression(random_state = 42).fit(x_train_scaled, y_train)

In [30]:
log_reg.predict(x_train_scaled)

array([1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1,
       1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0,
       1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0,
       1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0])

In [31]:
log_reg.score(x_train_scaled, y_train)

0.8660287081339713

In [33]:
log_reg.predict(x_test_scaled)

array([0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0,
       0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1,
       1, 1])

In [35]:
log_reg.score(x_test_scaled, y_test)

0.7444444444444445

In [38]:
print(data['AHD'].value_counts(normalize=True))


AHD
0    0.538462
1    0.461538
Name: proportion, dtype: float64


In [39]:
for c in [1.0, 0.5, 0.1, 0.01]:
    model = LogisticRegression(C=c, class_weight='balanced', max_iter=1000)
    model.fit(x_train_scaled, y_train)
    print(c, model.score(x_test_scaled, y_test))


1.0 0.7444444444444445
0.5 0.7444444444444445
0.1 0.7777777777777778
0.01 0.8


In [40]:
from sklearn.model_selection import cross_val_score
scores = cross_val_score(log_reg, x_train_scaled, y_train, cv=5)
print("Average CV Accuracy:", scores.mean())


Average CV Accuracy: 0.8233449477351916


In [43]:
log_reg = LogisticRegression(random_state = 42,
                             C = 0.01,
                             fit_intercept = True
                            ).fit(x_train_scaled, y_train)

In [44]:
log_reg.score(x_train_scaled, y_train)

0.8660287081339713

In [45]:
log_reg.score(x_test_scaled, y_test)

0.8111111111111111

Improved the model's accuracy significantly by adjusting regularization. I used a value of 0.01, although small, has a stronger regularization which generalizes better to unseen data and preventing overfitting. Test scores moved from 0.744 to 0.811, meaning my model is 81% accurate. Ont0 the next project!!!